crates/sui-types/src/intent.rs

```diff
@@ -0,0 +1,238 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::batch::AuthorityBatch;
+use crate::messages::{TransactionData, TransactionEffects};
+use crate::messages_checkpoint::CheckpointSummary;
+
+use serde::{Deserialize, Serialize};
+
+/// Current Sui version (for compatibility purposes).
+/// TODO: Increment value when a non-backwards compatible Sui version is launched (i.e. when
+/// changing the format, signature or field ordering of signable types).
+const SUI_COMPATIBILITY_VERSION: u8 = 0;
+const SUI_CHAIN_ID: u8 = 0;
+
+/// This is the only type we should sign per our serialization-handbook to provide domain separation
+/// and avoid accidental serialized-bytes collisions between structs and intents, different versions
+/// of the same struct and chains (mainnet, testnet etc).
+/// Similarly, `SecureIntent` should be preferred when hashing requires domain separation
+/// guarantees.
+pub trait SecureIntent: Serialize + serde::de::DeserializeOwned + private::SealedIntent {}
+
+/// Struct required when Sui repository-version and chainID are required for domain
+/// separation (i.e. when something is sign-able).
+#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
+pub struct VersionPlusChainId {
+    version: u8,
+    chain_id: u8,
```

Comment:
I'm not sure how I feel about chain_id being apart of the intent instead of say, part of the transaction itself. I suppose this is a good way to ensure everything i'm signing is for a particular chain. But it does fix our signature scheme to only really work with Sui. IIRC we want these intents to perhaps also be used for Narwhal and we may need some other domain separator for that case as chain_id may not be enough for our partners.

(I could be wrong about wanting to use this scheme for narwhal though)

Reply:
Agree with that and for a while I was planning to introduce an `app_id` as well to accommodate app-domain separation. If we do that, we have to maintain a list of apps and wanted to avoid that as external app devs (that they don't know about Sui at all) could easily violate that, unless we define a community standard where each app dev registers their app_id (as it happens with encoding schemes, see https://github.com/satoshilabs/slips/blob/master/slip-0173.md#registered-human-readable-parts).